plugins/processors/snmp_lookup/lookup.go

```diff
@@ -0,0 +1,341 @@
+	}
+
+	// Load from cache
+	tagMap, inCache := l.cache.Get(gs.Host())
+	tags, indexExists := tagMap.rows[index]
+	if inCache && indexExists {
+		for key, value := range tags {
+			metric.AddTag(key, value)
+		}
+	} else {
+		l.Log.Warnf("Could not find index %q on agent %q", index, gs.Host())
+	}
+
+	return []telegraf.Metric{metric}
+}
+
+// prepareCache prepares the cache if needed (index does not exist yet, or agent not yet cached)
+func (l *Lookup) prepareCache(gs snmpConnection, index string) error {
+	agent := gs.Host()
+
+	// Check cache
+	l.lock.Lock()
+	tagMap, inCache := l.cache.Peek(agent)
+	_, indexExists := tagMap.rows[index]
+
+	// Cache miss or non existing index and not recently refreshed the table
+	if !inCache || (!indexExists && time.Since(tagMap.created) > minRetry) {
+		// Check if another process is alreay loading the table, wait until done if so
+		if done, busy := l.sigs[agent]; busy {
+			l.lock.Unlock()
+			<-done
+		} else {
+			// No other process is already loading the table, let others know by creating a channel
+			l.sigs[agent] = make(chan struct{})
+			l.lock.Unlock()
+
+			if err := gs.Reconnect(); err != nil {
+				l.signalAgentReady(agent)
+				return fmt.Errorf("could not connect: %w", err)
+			}
+
+			// build the table for the configured tags
+			if err := l.loadTagMap(gs); err != nil {
+				l.signalAgentReady(agent)
+				return fmt.Errorf("could not load table: %w", err)
+			}
+
+			// Done loading, inform other processes
+			l.signalAgentReady(agent)
+		}
+	} else {
+		l.lock.Unlock()
+	}
+
+	return nil
+}
+
+func (l *Lookup) signalAgentReady(agent string) {
+	l.lock.Lock()
+	close(l.sigs[agent])
+	delete(l.sigs, agent)
+	l.lock.Unlock()
+}
+
+// getConnectionNoMock prepares a snmpConnection from the given metric tags (if present)
+func (l *Lookup) getConnectionNoMock(metric telegraf.Metric) (snmpConnection, error) {
+	clientConfig := l.ClientConfig
+
+	if version, ok := metric.GetTag(l.VersionTag); ok {
+		// inputs.snmp_trap reports like this
+		if version == "2c" {
+			version = "2"
+		}
+
+		v, err := strconv.ParseUint(version, 10, 8)
+		if err != nil {
+			return nil, fmt.Errorf("parsing version: %w", err)
+		}
+		clientConfig.Version = uint8(v)
+	}
+
+	if community, ok := metric.GetTag(l.CommunityTag); ok {
+		clientConfig.Community = community
+	}
+
+	if secName, ok := metric.GetTag(l.SecNameTag); ok {
+		clientConfig.SecName = secName
+	}
+
+	if secLevel, ok := metric.GetTag(l.SecLevelTag); ok {
+		clientConfig.SecLevel = secLevel
+	}
+
+	if authProtocol, ok := metric.GetTag(l.AuthProtocolTag); ok {
+		clientConfig.AuthProtocol = authProtocol
+	}
+
+	if authPassword, ok := metric.GetTag(l.AuthPasswordTag); ok {
+		clientConfig.AuthPassword = authPassword
+	}
+
+	if privProtocol, ok := metric.GetTag(l.PrivProtocolTag); ok {
+		clientConfig.PrivProtocol = privProtocol
+	}
+
+	if privPassword, ok := metric.GetTag(l.PrivPasswordTag); ok {
+		clientConfig.PrivPassword = privPassword
```

Comment:
I don't think it is a good idea to pass around credentials in tags as it is easy to leak those by forgetting to drop them before they do reach the output. I would really like to see all of those connection settings to be in the config, potentially in an extra-file that can be shared with the other SNMP plugins as I think they are required in all of them.

Reply:
I agree with this, will revert using tags for snmp credentials..

Reply:
Only problem is that the snmp_trap input puts this info as tags on the metric: https://github.com/influxdata/telegraf/tree/master/plugins/inputs/snmp_trap#metrics

So then it would not be possible to do a lookup for metrics coming from there?

Reply:
@srebhan ?

Reply:
Please?

Reply:
Would you agree to try still load config from the tags that are present from snmp_trap input?

Reply:
I really don't know... If the tags are already there it's not your fault to use them, but them being there is not a good thing... Could we get rid of those in the `snmp_trap`? If so I would not add them there but rather put an identifier for the credentials to use and those are defined in this plugin...

Not sure if this is understandable in any way... :-/

Reply:
Yes I understand.

Hmm, maybe we should drop this tag loading part anyway for the initial release of this plugin and add such functionality later when there is a generic method to define multiple snmp credentials in telegraf (so it can be used/reused by other plugins)

Reply:
Done in [8085bb8](https://github.com/influxdata/telegraf/pull/14223/commits/8085bb896058e8a24ab6aab9efe6c038af95be98)